File: src/ah/bundle.py
"""The world bundle (su-eng-01) — DN-3 W2's payload contract, versioned.

The browser never receives an ensemble: a bundle carries exactly what the
single-user experience renders, precomputed once and immutable (W10).
Contract v0.1, four sections plus provenance:

- ``revealed``   — the run's own path (monthly returns per asset, reported
                   variants for the smoothed sleeves) with its **t0 seal**
                   (``ah.artifacts.live.seal_tape`` over the numeric tape) so
                   live mode can prove nobody rewrote history mid-game (W3).
- ``bands``      — per-asset fan quantiles (p5/p25/p50/p75/p95 of cumulative
                   growth) drawn from the run's OWN regenerated ensemble.
- ``summary``    — final twin value, decision-window schedule, episode
                   markers (the world's regime sequence), digest verification.
- ``feed``       — the world's narrative dispatches, the store chronicle,
                   and (v0.2, su-app-03) ``artifacts``: the tier-1 wire
                   generated from the tape by ``ah.feed.build_tier1_feed``,
                   each item carrying its ``month`` so the app reveals it
                   with the pointer (E2). PD-4 stands: everything is authored
                   at BUILD time; tier-2 letters join only at the frozen
                   >=95% first-pass bar, and ``meta.artifact_tier`` records
                   what this bundle actually contains.

Determinism and lineage: the bundle is derived from the RunRecord ALONE via
the same regenerate-and-verify path as ``ah inspect`` — a tampered record
yields a loud ``digest_verified: false`` in the bundle rather than a pretty
payload. Two builds of the same run are byte-identical (no clocks; the only
timestamp is the record's own ``created_at``). Size budget: **under 1 MB
compressed**, enforced here (a bundle that outgrows W2's budget should fail
its build, not slowly poison load times).
"""

from __future__ import annotations

import gzip
import json
import sqlite3
from pathlib import Path
from typing import Any

import numpy as np

from ah.artifacts.live import seal_tape
from ah.core.digest import digest_ensemble
from ah.core.engine import ASSETS, REPORTED_SLEEVES, EnginePaths, run_ensemble, run_path
from ah.core.institution import decision_months, hold_course_twin
from ah.core.numericworld import project_numeric
from ah.core.worldspec import WorldSpec
from ah.feed import build_tier1_feed
from ah.play import simulate_play
from ah.store import chronicle as chronicle_store
from ah.store.runrecords import get_run_record
from ah.store.worlds import get_world

__all__ = ["BUNDLE_VERSION", "MAX_COMPRESSED_BYTES", "BundleError", "build_bundle", "write_bundle"]

BUNDLE_VERSION = "world-bundle-0.5"  # 0.5: factor lineage for generated worlds (su-gen-02)
MAX_COMPRESSED_BYTES = 1_000_000  # W2's budget: a complete world under 1 MB compressed
_QUANTILES = (5, 25, 50, 75, 95)

# The generated-world draw span, for proxy-share disclosure (survey S4: the
# per-factor share is reported over the months the sampler can actually draw)
_DRAW_SPAN = ("1953-04-01", "2020-12-01")


def factor_lineage(vintage_id: str) -> dict[str, Any]:
    """The 16-factor lineage block for a generated bundle: names' units and
    per-factor proxy shares (incl. the by-rule split the sealed
    ``proxy_share_disclosure`` clause requires — HAR separately from VXO).

    OD-4: this READS THE LOCAL CATALOG — building a generated bundle requires
    ``data/`` to be present, and the resulting numbers ship in the bundle so
    verification never needs the store. Raises :class:`BundleError` with the
    decision's name when the store is absent.
    """
    from ah.data.catalog import Catalog
    from ah.datalab import factor_read, proxy_share
    from ah.factors import load_manifest
    from ah.gen.bootstrap import FACTOR_SET

    root = Path(__file__).resolve().parents[2] / "data"
    if not (root / "catalog.duckdb").exists():
        raise BundleError(
            "generated bundles require the local vintage store (OD-4: build-time "
            f"data/ requirement with stamped lineage); {root} has no catalog"
        )
    cat = Catalog(root)
    manifest = load_manifest()
    units: dict[str, str] = {}
    shares: dict[str, Any] = {}
    for name in FACTOR_SET:
        fr = factor_read(cat, name, vintage=vintage_id)
        if fr.frame is None:
            raise BundleError(f"factor '{name}' unreadable on vintage {vintage_id}: {fr.note}")
        units[name] = manifest.sources[name].units or ""
        shares[name] = proxy_share(fr.frame, start=_DRAW_SPAN[0], end=_DRAW_SPAN[1])
    return {"vintage_id": vintage_id, "units": units, "proxy_shares": shares}


def _campaign_credibility(resolved_engine: dict[str, Any]) -> dict[str, Any]:
    """The audit-trail pointers a generated bundle carries: the campaign-3
    verdict (committed artifact, no store needed) plus the run's own lineage."""
    verdict_path = Path(__file__).resolve().parents[2] / "artifacts" / "campaign3"
    doc = json.loads((verdict_path / "promotion-verdict.json").read_text(encoding="utf-8"))
    return {
        "verdict": doc["verdict"],
        "campaign": doc["campaign"],
        "benchmark": doc["benchmark"],
        "severe": doc["severe"],
        "generator_id": resolved_engine.get("generator_id"),
        "generator_version": resolved_engine.get("generator_version"),
        "campaign_vintage_id": resolved_engine.get("campaign_vintage_id"),
    }


class BundleError(RuntimeError):
    """A bundle that cannot be built honestly (missing lineage, budget blown)."""


def _round(values: np.ndarray, places: int = 6) -> list[float]:
    return [round(float(v), places) for v in values]


def _twin_ledger(
    revealed: EnginePaths, start_targets: dict[str, float] | None = None
) -> dict[str, list[float] | list[int]]:
    """The hold-course institution's quarterly cashflows, for the bundle."""
    result = simulate_play(revealed, None, start_targets=start_targets)
    return {
        "quarter_months": [q.month for q in result.quarters],
        "calls": _round(np.array([q.calls_paid for q in result.quarters])),
        "distributions": _round(np.array([q.distributions_received for q in result.quarters])),
        "nav_true": _round(np.array([q.nav_true for q in result.quarters])),
        "nav_reported": _round(np.array([q.nav_reported for q in result.quarters])),
        "cash": _round(np.array([q.cash for q in result.quarters])),
        "unfunded": _round(np.array([q.unfunded_total for q in result.quarters])),
        "private_weight_true": _round(np.array([q.private_weight_true for q in result.quarters])),
    }


def build_bundle(conn: sqlite3.Connection, run_id: str) -> dict[str, Any]:
    """The W2 contract for one RunRecord, regenerated and verified from lineage."""
    rec = get_run_record(conn, run_id)
    if rec is None:
        raise BundleError(f"no run_record with run_id={run_id}")
    world_doc = get_world(conn, rec["world_id"])
    if world_doc is None:
        raise BundleError(f"run {run_id} references missing world {rec['world_id']}")

    ws = WorldSpec.model_validate(world_doc)
    nw = project_numeric(ws)
    generated = ws.engine_defaults.generator_id != "toy-v0"
    if generated:
        from ah.port.adapter import (
            GEN_START_MIX,
            GEN_START_TARGETS,
            gen_hold_course_twin,
            run_gen_ensemble,
            run_gen_path,
        )

        revealed = run_gen_path(nw, rec["seed"])
        ensemble = run_gen_ensemble(nw, rec["n_paths"], base_seed=rec["seed"])
        twin = gen_hold_course_twin(nw, rec["seed"])
        play_targets: dict[str, float] | None = GEN_START_TARGETS
        feed_kwargs: dict[str, Any] = {"run_path_fn": run_gen_path, "start_mix": GEN_START_MIX}
    else:
        revealed = run_path(nw, rec["seed"])
        ensemble = run_ensemble(nw, rec["n_paths"], base_seed=rec["seed"])
        twin = hold_course_twin(nw, rec["seed"])
        play_targets = None
        feed_kwargs = {}
    verified = digest_ensemble(ensemble) == rec["outputs_digest"]

    # The numeric tape (months x series): asset returns then reported columns,
    # in the tape's own recorded order — the t0 seal covers exactly these bytes.
    asset_order = revealed.asset_order
    series_order = [*asset_order, *(f"{s}_reported" for s in REPORTED_SLEEVES)]
    tape = np.column_stack(
        [revealed.returns[a] for a in asset_order]
        + [revealed.reported[s] for s in REPORTED_SLEEVES]
    ).astype(np.float64)
    # The seal covers the bytes the bundle SHIPS: round first, seal the rounded
    # tape, store the same values -- a client re-sealing what it received must
    # reproduce the hash exactly.
    tape = np.round(tape, 6)

    bands: dict[str, dict[str, list[float]]] = {}
    for asset in ensemble.asset_order:
        # engine returns are in PERCENT (see ah.core.engine notes) — divide
        # before compounding, or the cone explodes/goes negative (found live:
        # the first played decade rendered empty fans at the broken scale)
        growth = np.cumprod(1.0 + ensemble.returns[asset] / 100.0, axis=1)
        qs = np.percentile(growth, _QUANTILES, axis=0)
        bands[asset] = {f"p{q}": _round(qs[i]) for i, q in enumerate(_QUANTILES)}

    narrative = world_doc.get("narrative") or {}
    regimes = world_doc.get("regimes") or {}
    chronicle_rows = chronicle_store.read(conn, rec["world_id"])

    doc: dict[str, Any] = {
        "bundle_version": BUNDLE_VERSION,
        "meta": {
            "world_id": rec["world_id"],
            "run_id": run_id,
            "seed": rec["seed"],
            "n_paths": rec["n_paths"],
            "months": revealed.months,
            "created_at": rec["created_at"],
            "resolved_engine": rec["resolved_engine"],
            "decision_stamps": {
                "decision_schema_version": rec.get("decision_schema_version"),
                "decision_alpha_version": rec.get("decision_alpha_version"),
                "twin_definition": rec.get("twin_definition"),
            },
            "artifact_tier": (
                "tier-1 templated wire (build-time, deterministic); "
                "tier-2 letters await the frozen >=95% first-pass bar"
            ),
            "digest_verified": bool(verified),
            "outputs_digest": rec["outputs_digest"],
            "title": narrative.get("title"),
            "tagline": narrative.get("tagline"),
        },
        "revealed": {
            "series_order": series_order,
            "tape": [_round(row) for row in tape],
            "tape_seal": seal_tape(tape),
        },
        "bands": bands,
        # The HOLD-COURSE TWIN's cashflows. Decision-independent by
        # construction — the twin never acts — so it stays pre-authorable at
        # build time (PD-4). The player's own ledger depends on their
        # decisions and is served per-session instead.
        "twin_ledger": _twin_ledger(revealed, play_targets),
        "summary": {
            "twin_final_value": float(twin.final_value),
            "decision_months": decision_months(revealed.months),
            "episodes": regimes.get("sequence") or [],
            "summary_stats": rec["summary_stats"],
        },
        "feed": {
            "artifacts": build_tier1_feed(
                nw, revealed, base_seed=rec["seed"], n_peer_paths=rec["n_paths"], **feed_kwargs
            ),
            "dispatches": narrative.get("dispatches") or [],
            "chronicle": [
                {
                    "seq": row["seq"],
                    "type": row["type"],
                    "run_id": row.get("run_id"),
                    "created_at": row["created_at"],
                }
                for row in chronicle_rows
            ],
        },
    }
    if generated:
        # su-gen-02: the factor lineage + the audit trail. Both are static
        # data stamped at build; their integrity rides in meta.sections_seal
        # (the tape_seal covers the tape alone).
        import hashlib

        from ah.core.digest import canonical_json

        doc["factors"] = {
            "names": _factor_names(),
            **factor_lineage(str(rec["resolved_engine"].get("campaign_vintage_id"))),
        }
        doc["credibility"] = _campaign_credibility(rec["resolved_engine"])
        extras = canonical_json({"factors": doc["factors"], "credibility": doc["credibility"]})
        doc["meta"]["sections_seal"] = (
            "sha256:" + hashlib.sha256(extras.encode("utf-8")).hexdigest()
        )
    return doc


def _factor_names() -> list[str]:
    from ah.gen.bootstrap import FACTOR_SET

    return list(FACTOR_SET)


def write_bundle(doc: dict[str, Any], path) -> int:
    """Serialize + gzip; enforce W2's size budget. Returns compressed bytes."""
    from pathlib import Path

    payload = json.dumps(doc, sort_keys=True, separators=(",", ":")).encode("utf-8")
    compressed = gzip.compress(payload, compresslevel=9, mtime=0)  # mtime=0: determinism
    if len(compressed) > MAX_COMPRESSED_BYTES:
        raise BundleError(
            f"bundle is {len(compressed)} bytes compressed, over W2's "
            f"{MAX_COMPRESSED_BYTES}-byte budget — shrink it, do not ship it"
        )
    target = Path(path)
    target.parent.mkdir(parents=True, exist_ok=True)
    target.write_bytes(compressed)
    return len(compressed)

File: app/src/lib/bundle.ts
/**
 * World-bundle loading + verification (su-app-01; DN-3 W2/W3).
 *
 * The bundle is the ONLY payload the browser consumes (never an ensemble).
 * Loading is three steps, all client-side:
 *   1. fetch the .bundle.gz and gunzip it (DecompressionStream — no deps)
 *   2. parse and shape-check the W2 contract sections
 *   3. re-seal the revealed tape locally and compare with `tape_seal` —
 *      the client PROVES nobody rewrote history, it does not take it on faith.
 *
 * The seal algorithm mirrors ah.core.digest.sha256_of_arrays: SHA-256 over
 * the row-major float64 bytes of the (months x series) tape, little-endian.
 * The server rounds to 6dp before sealing, so the JSON numbers we received
 * ARE the sealed values — reconstructing a Float64Array from them and
 * hashing reproduces the seal exactly.
 */

export const SUPPORTED_BUNDLE_VERSIONS = [
  "world-bundle-0.1",
  "world-bundle-0.2",
  // 0.3 added `private`: the display-only pacing ledger. Older bundles still
  // load — the panel simply has nothing to draw.
  "world-bundle-0.3",
  // 0.4 replaces `private` with `twin_ledger`: the hold-course twin's
  // quarterly cashflows, now that the play surface runs on the real
  // institutional twin instead of a toy target-mix simulator.
  "world-bundle-0.4",
  // 0.5 (su-gen-02): generated worlds carry `factors` (the 16 factor names,
  // units, per-factor proxy shares) and `credibility` (the campaign-3
  // verdict + vintage - the audit trail behind "rearranged truth"). Toy
  // bundles are unchanged in shape; the two sections are optional.
  "world-bundle-0.5",
  // 0.6 (ER-14 close-out, D-ER14-2): the fourth private class (infra) joins
  // the played book -- thirteen toy series, twelve generated. No new
  // top-level section; every series this bundle carries flows through the
  // same `revealed`/`bands` shapes already handled below.
  "world-bundle-0.6",
];

/** su-gen-02: factor lineage for generated worlds (optional section). */
export interface BundleFactors {
  names: string[];
  vintage_id: string;
  units: Record<string, string>;
  proxy_shares: Record<
    string,
    { n_months: number; n_proxy: number; share: number; by_rule: Record<string, number> }
  >;
}

/** su-gen-02: the campaign audit trail (optional section). */
export interface BundleCredibility {
  verdict: string;
  campaign: string;
  benchmark: string;
  severe: Record<string, unknown>;
  generator_id: string | null;
  generator_version: string | null;
  campaign_vintage_id: string | null;
}

export interface BundleMeta {
  world_id: string;
  run_id: string;
  seed: number;
  n_paths: number;
  months: number;
  created_at: string;
  digest_verified: boolean;
  outputs_digest: string;
  artifact_tier: string;
  title: string | null;
  tagline: string | null;
  decision_stamps: Record<string, string | null>;
  resolved_engine: Record<string, string>;
}

/** One tier-1 wire item (bundle v0.2); revealed with the pointer (E2). */
export interface FeedArtifact {
  month: number;
  type: string;
  payload: {
    dateline: string;
    title?: string;
    headline?: string;
    lines?: string[];
    release_name?: string;
    rows?: { series: string; value: string; prior: string; revision: string }[];
    /** sp-04: board packs carry titled sections */
    sections?: { title: string; lines: string[] }[];
  };
}

/** The hold-course twin's quarterly cashflows, carried in the bundle.
 *  Decision-independent by construction, which is why it can be
 *  pre-authored; the player's own ledger comes from the session. */
export interface TwinLedger {
  quarter_months: number[];
  calls: number[];
  distributions: number[];
  nav_true: number[];
  nav_reported: number[];
  cash: number[];
  unfunded: number[];
  private_weight_true: number[];
}

export interface WorldBundle {
  bundle_version: string;
  meta: BundleMeta;
  revealed: { series_order: string[]; tape: number[][]; tape_seal: string };
  bands: Record<string, Record<string, number[]>>;
  /** present from world-bundle-0.4; absent in older bundles */
  twin_ledger?: TwinLedger;
  factors?: BundleFactors;
  credibility?: BundleCredibility;
  summary: {
    twin_final_value: number;
    decision_months: number[];
    episodes: unknown[];
    summary_stats: Record<string, number>;
  };
  feed: { artifacts?: FeedArtifact[]; dispatches: unknown[]; chronicle: unknown[] };
}

export class BundleFormatError extends Error {}

/** SHA-256 over the tape's row-major float64 bytes — the seal's exact recipe. */
export async function sealTape(tape: number[][]): Promise<string> {
  const rows = tape.length;
  const cols = rows > 0 ? tape[0].length : 0;
  const flat = new Float64Array(rows * cols);
  for (let r = 0; r < rows; r++) {
    const row = tape[r];
    if (row.length !== cols) throw new BundleFormatError("ragged tape");
    for (let c = 0; c < cols; c++) flat[r * cols + c] = row[c];
  }
  const digest = await crypto.subtle.digest("SHA-256", flat.buffer as ArrayBuffer);
  const hex = [...new Uint8Array(digest)].map((b) => b.toString(16).padStart(2, "0")).join("");
  return `sha256:${hex}`;
}

function shapeCheck(doc: WorldBundle): void {
  if (!SUPPORTED_BUNDLE_VERSIONS.includes(doc.bundle_version)) {
    throw new BundleFormatError(
      `unsupported bundle_version ${doc.bundle_version} (want one of ${SUPPORTED_BUNDLE_VERSIONS.join(", ")})`,
    );
  }
  for (const section of ["meta", "revealed", "bands", "summary", "feed"] as const) {
    if (!(section in doc)) throw new BundleFormatError(`missing section ${section}`);
  }
  const { months } = doc.meta;
  if (doc.revealed.tape.length !== months) {
    throw new BundleFormatError(
      `tape has ${doc.revealed.tape.length} months, meta says ${months}`,
    );
  }
}

export interface LoadedBundle {
  bundle: WorldBundle;
  /** The client's own re-seal matched the shipped seal. */
  sealVerified: boolean;
}

export async function parseBundle(gzBytes: ArrayBuffer): Promise<LoadedBundle> {
  const stream = new Blob([gzBytes]).stream().pipeThrough(new DecompressionStream("gzip"));
  const json = await new Response(stream).text();
  const bundle = JSON.parse(json) as WorldBundle;
  shapeCheck(bundle);
  const sealVerified = (await sealTape(bundle.revealed.tape)) === bundle.revealed.tape_seal;
  return { bundle, sealVerified };
}

export async function fetchBundle(url: string): Promise<LoadedBundle> {
  const res = await fetch(url);
  if (!res.ok) throw new BundleFormatError(`fetch ${url}: ${res.status}`);
  return parseBundle(await res.arrayBuffer());
}
